### ISSUE TITLE
There is no such thing as implicitly agreeing to relicense

### ISSUE THREAD

**Issue by nurupo** [lerna/lerna#1620]
@kittens @jamiebuilds @evocateur I see you guys saying

>[Even still, all contributors implicitly agreed to the existing license, of which I am the original license holder, when they submitted code meaning we are within our rights to relicense.](https://github.com/lerna/lerna/pull/1616#issuecomment-416820399)

Did you somehow miss the shitstorm that OpenSSL relicensing was a year ago, with their

>[We are asking for your permission to change the licence for your
contribution. [...] If we do not hear from you, we will assume that you have no objection.](https://lwn.net/Articles/717996/)

?

After which OpenSSL came to their senses and removed that "we will assume that you have no objection" part, [seeking explicit agreement from contributors and removing their code if they didn't agree](https://www.openssl.org/blog/blog/2017/06/17/code-removal/).

It's not just OpenSSL, VLC [[1]](http://blog.klocwork.com/open-source/vlc-license-change-a-lesson-in-perseverance/) [[2]](http://www.jbkempf.com/blog/post/2012/How-to-properly-relicense-a-large-open-source-project), ZNC [[3]](https://github.com/znc/znc/issues/311) had to reach out to every contributor in order to relicense too.

Long story short, you have to get an explicit permission from every contributor to relicense their code, if they don't agree or don't contact you back -- you can't relicense it, you can either keep it under the initial license or drop their code. There is no implicit agreeing. Implicit agreement could be declared in a CLA every contributor has to sign, but I see no CLA in lerna.

**Comment by nickmccurdy** [comment]
Even though sublicensing is allowed, because these projects have no contributor license agreements and every contributor did not agree, the licenses are all invalid. I am not a lawyer and this is not legal advice, but I suggest changing all your licenses to original MIT, or you may be open to lawsuits from prior contributors (including companies on the list).

**Comment by pluma** [comment]
@nickmccurdy technically (which the link backs up) a CLA is necessary for any contribution in order to be able to distribute that contribution under any license at all (because without a CLA the contribution itself comes with no license, i.e. "all rights reserved"). Most community maintained open source projects simply assume contributors implicitly agree to the terms of the license of the project they're contributing though:

> [CLAs] are a way for the contributor to tell the project that it has the right to distribute the new contributions under the project's **existing** open source license

(emphasis added)

So IOW every open source project technically needs CLAs but that's irrelevant to the relincensing question as long as the relicensing is permitted by the existing license. MIT is generally held to permit relicensing under a more restrictive license, which "MIT with a blacklist" certainly is.

So if anything you could argue that the contributions to Lerna by external contributors are all unlicensed and were never legal to be distributed under the MIT license because there was no CLA or CAA. But that argument can be made for the vast majority of open source projects that aren't backed by companies with legal departments.

If we accept that the distribution of the contributions under the MIT license was legitimate without a CLA, the license change is not a problem. If we don't accept that, the original license was already invalid (but that would be true for 99.9% of GitHub projects that accept PRs).

**Comment by pluma** [comment]
Well, again, IANAL, but as I understand it nothing can protect you against being sued or threatened with legal actions. The question is if a prior contributor might succeed if they sued but that's difficult to predict.

As a non-lawyer I'd say that redistributing contributions submitted as PRs on GitHub under the same license stated on that same repository is standard practice and submitting a PR therefore means you implicitly agree to having your contributions covered by that license.

That said, if a contributor were to sue over this and succeed, they'd set a legal precedent that would require all open source projects to adopt CLAs to prevent being sued by their contributors, regardless of their licenses or planned licensing changes.

**Comment by pluma** [comment]
FYI the licensing changes originally cited as examples were

VLC: GPLv2 -> LGPLv2.1
ZNC: GPLv2 -> Apache 2.0

In both cases the new license is more permissive, meaning the authors grant the users additional rights. In the case of the Apache 2.0 license there's also a patent grant involved, which causes additional headaches.

The license change lerna just implemented merely restricted who the license applies to (i.e. not any of the companies on the blacklist) and reserves all rights to everyone else. So it's from permissive to less permissive (or equally permissive but to fewer people). That seems less problematic.

**Comment by MadaraUchiha** [comment]
What do you mean it's less problematic?

If I were to set the license to "Only those that paid @MadaraUchiha money" it's also most restrictive and no additional rights are given, how is that any less problematic?

**Comment by pluma** [comment]
It's less problematic because you're allowed to do that by the MIT license.

The MIT license explicitly grants the right to sublicense and even to sell copies of the software (without also providing access to the source code as the GPL would require).

Lerna's license could be changed to

> Permission is hereby granted, ~~free of charge~~ _in exchange for one million dollars paid to @MadaraUchiha per individual use_, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

... and that'd be entirely within the scope of what the MIT license allows you to do.

Keep in mind that license changes aren't retroactive. A copy distributed under the MIT license remains under that license. Making a copy of it and changing to a more restrictive license is legal. Making a copy of it and keeping the license is also legal. But making a copy of the copy that has a more restrictive license and re-releasing it under the old license wouldn't be legal.

Imagine licenses as a sieve for rights held by the authors. If you slap a permissive sieve on, more rights are granted to the user. If you slap a restrictive sieve on, less rights are granted. If you slap a more restrictive sieve on top of a more permissive sieve, less rights are granted than before. If you slap a more permissive sieve on top of a more restrictive sieve it won't grant more rights so that doesn't work. The most restrictive sieve is the default ("all rights reserved"), the least restrictive sieve is the public domain. Some sieves have spikes that prevent slapping most other sieves on top of them (e.g. the GPL). ... Okay, this analogy isn't perfect but I think you get the idea.

**Comment by MadaraUchiha** [comment]
@pluma Sure, but again, only with the agreement (or contribution removal) of every other collaborator. That's the whole dispute, that there's no such thing as "implicit" agreement to this.

**Comment by pluma** [comment]
Those two things are the same thing. It's not retroactive and it's not granting additional rights the contributors didn't agree to. For all intents and purposes a relicensing is a fork that retains the same name and location.

**Comment by sdkuck** [comment]
Interesting point that you can't move to a more permissive license without permission. That means lerna will be required to get permission from every contributor from now on if they want to remove these restrictions from the license in the future.

**Comment by MadaraUchiha** [comment]
@sdkuck Only those contributors who gave contributions under the new license. Those who made their contributions under the original license already gave their permission to release their contribution under the original license, and so you wouldn't need to ask it from them again.

**Comment by dtibarra** [comment]
Code contributions only receive the the MIT license once the code is at rest in the repo after merge. Contributions themselves are not MIT licensed, even though they are a derivative work because the act of applying means only receiving and merging the code diff in question, not code diff plus MIT copyright notice and permission notice as stated in the terms of the MIT license

The terms of the MIT license state, "Permission is hereby granted, free of charge, to any person obtaining **a copy of this software** and associated documentation files"

I'd argue that this codebase is not a copy, but the original. I'd also argue that anyone who does not agree to this sublicense has legal recourse due to a violation of their copyright.

**Comment by kittens** [comment]
Just to clarify, while I would not personally have chosen to go with the new sublicense, I do support the existing maintainers who have completely rewritten Lerna from what t was when I first started the project. I do not consider this project to be mine.

Like the others have explained, this is making the license more restrictive, not submissive which makes the license change completely valid.

This is very different to the OpenSSL case where the entire license was changed. As far as I can tell, they didn’t even need my permission to make this change.

**Comment by pluma** [comment]
@dtibarra I don't think that interpretation is correct. You're talking about software being developed using a distributed version control system. The version "hosted" on GitHub right now is just a working copy of the latest commit pushed to master. There's no "original" as everything is a copy. The license also makes no mention of the "original" as it doesn't make sense for software code. If anything, the "original" of lerna lives on the harddisk of its first original author.

But even if you were right, what's the practical difference between relicensing "the original" versus making a copy of the original under a new license, moving the original somewhere else and replacing the original with the copy?

The "original" lives here and it's currently at version 3.2.1: https://github.com/lerna/lerna/tree/v3.2.1 -- it's still under MIT and always will be, just as it always will be at version 3.2.1. If you want to continue work on the original, you can fork version 3.2.1 and release your own versions under whatever license you want (as long as it's compatible with MIT -- e.g. you can't put it in the public domain).

If you want to sue, you can try revoking the usage rights you originally implicitly granted as part of the PR (but you'll probably have better luck claiming you never knowingly granted them in the first place). But I've never heard of any precedent for this except in cases where the contributor (knowingly or not) didn't actually own the code they contributed.

In all likelihood however I doubt your individual contribution is significant enough that you couldn't simply ask for your PR to be reverted and the maintainers would comply out of courtesy -- if you haven't burned that bridge by grabbing tikki torches and pitchforks first.

**Comment by orthecreedence** [comment]
> Also note that license changes aren't retroactive.

@pluma Exactly. You cannot relicense code without a) a previous agreement that allows a relicense (CLA) or b) explicit permission from every contributor c) a clause in the license that explicitely allows relicensing by the original copyright holder.

With copyright law, the people contributing the code own the code, unless they sign that right away explicitely. If they contribute it under the terms of a specific license (MIT), they still own that code. MIT doesn't transfer ownership, and it doesn't allow explicit relicensing. So if the maintainers want to relicense as MIT-ICE they need permission.

Even if they fork the project, they need permission to change the terms under which the code is distributed. This is why people set up CLAs, so they can relicense later without having to ask permission.

**Comment by jamiebuilds** [comment]
I think @kittens has resolved this issue
